controller/backing_image_controller.go

```diff
@@ -310,22 +345,604 @@ func (bic *BackingImageController) syncBackingImage(key string) (err error) {
+		}
+
+		instanceManager, err := bic.ds.GetRunningInstanceManagerByNodeRO(node.Name, longhorn.DataEngineTypeV2)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get the default instance manager for node %v", node.Name)
+		}
+		engineClientProxy, err := engineapi.NewEngineClientProxy(instanceManager, log, bic.proxyConnCounter)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get the engine client proxy for instance manager %v", instanceManager.Name)
+		}
+
+		log.Infof("Deleting the failed v2 copy on disk %v", v2DiskUUID)
+		err = engineClientProxy.SPDKBackingImageDelete(bi.Name, v2DiskUUID)
+		engineClientProxy.Close()
+		if err != nil {
+			return errors.Wrapf(err, "failed to delete the v2 copy on diskUUID %v", v2DiskUUID)
+		}
+
+		delete(bi.Status.DiskFileStatusMap, v2DiskUUID)
+		bic.v2CopyBackoff.DeleteEntry(v2DiskUUID)
+		bic.eventRecorder.Eventf(bi, corev1.EventTypeNormal, constant.EventReasonDelete, "Deleted failed v2 backing image %v on disk %v on node %v", bi.Name, v2DiskUUID, node.Name)
+	}
+	return nil
+}
+
+func (bic *BackingImageController) cleanupAllV2BackingImageCopies(bi *longhorn.BackingImage) (cleaned bool, err error) {
+	log := getLoggerForBackingImage(bic.logger, bi)
+
+	if bic.syncV2StatusWithInstanceManager(bi); err != nil {
+		return false, errors.Wrapf(err, "failed to sync v2 backing image status from instance manager")
+	}
+
+	// we can delete the backing image when all the v2 copies are deleted
+	if len(bi.Status.DiskFileStatusMap) == 0 {
+		return true, nil
+	}
+
+	allV2CopiesCleaned := true
+	for diskUUID := range bi.Status.DiskFileStatusMap {
+		if fileSpec, exists := bi.Spec.DiskFileSpecMap[diskUUID]; exists && fileSpec.DataEngine != longhorn.DataEngineTypeV2 {
+			continue
+		}
+
+		allV2CopiesCleaned = false
+
+		node, _, err := bic.ds.GetReadyDiskNodeRO(diskUUID)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to get the ready disk node for disk %v", diskUUID)
+		}
+
+		instanceManager, err := bic.ds.GetRunningInstanceManagerByNodeRO(node.Name, longhorn.DataEngineTypeV2)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to get the default instance manager for node %v", node.Name)
+		}
+		engineClientProxy, err := engineapi.NewEngineClientProxy(instanceManager, log, bic.proxyConnCounter)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to get the engine client proxy for instance manager %v", instanceManager.Name)
+		}
+
+		log.Infof("Deleting the v2 copy on disk %v for backing image %v", diskUUID, bi.Name)
+		err = engineClientProxy.SPDKBackingImageDelete(bi.Name, diskUUID)
+		engineClientProxy.Close()
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to delete the v2 backing image on diskUUID %v", diskUUID)
+		}
+
+		delete(bi.Status.DiskFileStatusMap, diskUUID)
+		bic.eventRecorder.Eventf(bi, corev1.EventTypeNormal, constant.EventReasonDelete, "Deleted v2 backing image %v on disk %v on node %v", bi.Name, diskUUID, node.Name)
```

Comment:
Suggest abstracting the codes into a function and close the client in a defer function.

Reply:
fixed.